Remove capnproto

dkeeney opened the pull request:
This is still not expected to compile.  There are still some errors in the Swig generated code I am trying to resolve.

dkeeney commented:
I still have one problem:  Swig:  "Error: Nothing known about namespace 'nupic::algorithms::connections'
while compiling algorithms.i.  It does have %include <nupic::algorithms::Connections.hpp> which defines the namespace 'nupic::algorithms::connections' but Swig is still very confused about something.

I don't really want to spend very much more time debugging code (the SWIG bindings) which we will be deleting entirely.  Can someone take a quick look and see if they can find the problem. If not, shall I just go on to the next PR and delete SWIG?  This does compile on Windows without SWIG.

dkeeney commented on src/nupic/algorithms/Cells4.cpp:
```diff
@@ -2183,16 +1954,14 @@ void Cells4::save(std::ostream &outStream) const {
+//do later  std::filesystem::create_directories(p.parent_path());
+  std::ofstream out(filePath.c_str(),std::ios_base::out | std::ios_base::binary);
+  out.exceptions(std::ofstream::failbit | std::ofstream::badbit);
+  out.precision(std::numeric_limits<double>::digits10 + 1);
```
I have no idea what the +1 is for.   All of the examples showed using it
f << std::setprecision(std::numeric_limits<long double>::digits10 + 1);
The docs don't say what it is for.

dkeeney commented on src/nupic/algorithms/Cells4.hpp:
```diff
@@ -637,510 +645,513 @@ class Cells4 : public Serializable<Cells4Proto> {
+        {
+          // TODO: this won't scale!
+          std::stringstream tmp;
+          this->save(tmp);
```
Don't think so.  Just 64 bit address space.  It would swap onto disk if it got too big but it would still work.

dkeeney commented on src/nupic/types/Serializable.hpp:
```diff
@@ -22,48 +22,72 @@
 /** @file
  * Definitions for the base Serializable class in C++
+ *
+ * CapnProto serialization has been removed and replaced with binary streams.
```
Good idea

dkeeney commented:
I have this branch building again.  The problem was that I was too aggressive when ripping things out of algorithm.i and got a few more lines that I intended.  That was a tough one to debug.
I will upload it as soon as I address the rest of breznak's comments on the last one.

dkeeney commented:
Another design question.  We have
- serialization to binary   (the save() and load())
- serialization to text
- toString()
Which one should the << and >> overloading apply to.  I am thinking it should be serialization to text. Does anyone care?

dkeeney commented on src/nupic/types/Serializable.hpp:
```diff
@@ -22,48 +22,72 @@
-#include <capnp/message.h>
-#include <capnp/serialize.h>
-#include <kj/std/iostream.h>
+#define SERIALIZABLE_VERSION 1
```
There is a problem with doing that...some of the things being serialized depend on the version of the object not the version of the overall Serialization protocol.

dkeeney commented on src/nupic/types/Serializable.hpp:
```diff
@@ -22,48 +22,72 @@
-#endif // NTA_serializable_HPP
+
+/*****
+using namespace nupic::serializable;
```
That is what I was thinking.  However it means that anyone trying to use this must also include yet another namespace.  The objective of a namespace is to avoid name collisions.  Since the Serializable class can appear anywhere within the nupic namespace and save() and load() should not mean anything else, it seems reasonable to include it in the nupic namespace.  So I am leaving out the serializable namespace.

dkeeney commented on src/test/unit/algorithms/Cells4Test.cpp:
```diff
@@ -179,94 +175,7 @@ TEST(Cells4Test, pickleSerialization) {
 }
 
-TEST(Cells4Test, capnpSerialization) {
-  Cells4 cells(10, 2, 1, 1, 1, 1, 0.5, 0.8, 1, 0.1, 0.1, 0, false, -1, true,
```
Note the test just before this one is pickleSerialization. This is the same test but using binary serialization.  I renamed it just Serialization because it is pickling only in the Python.

dkeeney commented on src/nupic/types/Serializable.hpp:
```diff
@@ -22,48 +22,72 @@
-  virtual void write(typename ProtoT::Builder &proto) const = 0;
-  virtual void read(typename ProtoT::Reader &proto) = 0;
+  // These must be implemented by the subclass.
+  virtual void save(std::ostream &stream) const = 0;
```
Fixed.  Good catch

breznak commented:
Dave, great you've managed to fix the stuff in Algorithms,i ! 
I'm at a conference this week, so apologies for slower replies.. 

> Which one should the << and >> overloading apply to. I am thinking it should be serialization to text. Does anyone care? 

by "serialization to text" you mean a nicer print()? ie python's `__str__` meaning? Otherwise, if there were 2 full-blown serializations, I'd go with just one (textual)

breznak commented on src/nupic/algorithms/Svm.hpp:
```diff
@@ -325,7 +326,8 @@ struct decision_function {
  * probA, probB = [n_class*(n_class-1)/2]
  */
-class svm_model {
+class svm_model  : public Serializable
```
btw, we should investigate if we can get rid off this file

breznak commented on src/nupic/engine/Link.cpp:
```diff
@@ -476,7 +474,7 @@ void Link::serialize(std::ostream &f) {
     f << a; // our part of the current Dest Input buffer.
 
-    std::deque<Array>::iterator itr;
+    boost::circular_buffer<Array>::iterator itr;
```
why this change from std:: ?

dkeeney replied:
I started to make that a std::deque but discovered that it breaks other code that should be in a different PR.  It's part of a PR to remove dimensions which ends up removing LinkPolicy and removes a lot of complexity in the Link object.  So lets delay that for now.

breznak commented on src/nupic/math/Array2D.hpp:
```diff
@@ -36,6 +36,8 @@
 // try to separate algorithms and write them for all types of matrices??
 // optimize kernels depending on cache sizes?
 
+// Note: cannot be made a nupic::Serializable class because it is not in nupic namespace.
```
can you add it to ns nupic:: ?

dkeeney replied:
Well...perhaps but lets make that a separate PR.  It could get messy.

breznak commented on src/nupic/math/ArrayAlgo.hpp:
```diff
@@ -24,4952 +24,6021 @@
+ *	bind2nd() is depriciated in C++11 and removed in C++17 so
+ *            bind2nd is replaced with C++11 bind(). At some point we may
+ *            want to consider using C++14 generic lambda which they say is faster.
+ *  Compiles under Windows C++17 and Linux C++11.  Asm no used in Windows.
```
the comment and change to bind() is :+1: 
but this file is probably a merge conflict! Too many changes here, 
please see the original 
https://github.com/htm-community/nupic.cpp/blob/master/src/nupic/math/ArrayAlgo.hpp 
And apply your change to that. 
I think you've re-added the ASM stuff here (which might be the reason why CI builds are failing) 
((also, this removed ASM is another thing to add to the "Changes", although it's not API(?) )

dkeeney replied:
This is a tough one.  I am trying to find something that will compile under C++11 and C++17.  I ended up getting really messed up such that nothing compiled.  So I went back to the original...perhaps I went back too far.  It now compiles.  I don't think the ASM stuff is all that bad. This was not the problem with CI.  It will not work on Windows but in those cases where it matters it does make a big difference in speed.   But I will work on this some more and reduce the conflicts.

breznak replied:
> compile under C++11 and C++17 

for other PR, but.. you've convinced me to go c++17 way (but there were some problems with python you've later discovered?), so if there are problems between ++11/17, should we move to the middleman c++14 as a sub-goal? 

> where it matters it does make a big difference in speed 

Do you have a test for this? Because I was deciding just that, and in my evaluation the plain c++(with compiler optimizations) made no difference compared to the ASM - therefore I went on pruning all of the ASM. 

Anyway, my comment was that you got messed in rebasing from your other branch to the newer base, if there's not too much changes, I'd suggest go with the current file from this repo and rewrite your changes there manually.

dkeeney replied:
Ok, re-did this file.  Went back to nupic.cpp master, and re-made just my changes. .. not many.  Compiles on C++11 and C++17.

breznak commented on src/nupic/math/StlIo.hpp:
```diff
@@ -551,37 +551,48 @@ inline std::ostream &operator<<(std::ostream &out_stream,
 
 //--------------------------------------------------------------------------------
 // std::map
+// Warning: This will not handle elements containing whitespace.
```
can we NTA_ASSERT on this?

dkeeney replied:
I am not sure this is actually used anywhere...just noticed this was the case while looking to add something like it for std::deque.   The way to allow whitespace would be to use readline and terminate with \n.  We can keep that in mind if we find we are using this function.

breznak commented on src/nupic/ntypes/ArrayBase.cpp:
```diff
@@ -391,138 +392,7 @@ void ArrayBase::binaryLoad(std::istream &inStream) {
-    out << YAML::Key << "count" << YAML::Value << count_;
-    out << YAML::Key << "buffer" << YAML::Value << YAML::BeginSeq;
-    switch (type_) {
-    case NTA_BasicType_Byte:
```
nice cleanup!

dkeeney replied:
:)

dkeeney commented on API_CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+* CapnProto serialization is replaced with binary streams in PR #62.  
+Calls to read() and write() are no longer available. Use save() and load(). Network(path) is no 
+longer used to deserialize a path. Use Network net; net.load(stream); to deserialize.  
+Helpers SaveToFile(path) and LoadFromFile(path) are used to stream to and from a file using save() 
```
ok

dkeeney commented on src/nupic/algorithms/Cells4.cpp:
```diff
@@ -1880,206 +1892,373 @@ void Cells4::reset() {
   //}
 }
 
+struct self_t {
```
Oh, I forgot about compiler generated padding.  Good catch.
I did zero fill the structure before filling it so any padding will be nulls.

dkeeney commented on src/nupic/algorithms/Cells4.cpp:
```diff
@@ -1880,206 +1892,373 @@ void Cells4::reset() {
+  UInt pamCounter;
+};
+// NOTE:  ownsMemory cannot be saved in serialization.
+//        It indicates who currently allocated the buffers.
```
Well...I need this for CSharp too because CSharp has a different heap (managed) than C++.  It cannot directly access C++'s heap space.  So to avoid a copy we need to do this.  I will try to make this a little cleaner however.

dkeeney commented on src/nupic/algorithms/Cells4.cpp:
```diff
@@ -1880,206 +1892,373 @@ void Cells4::reset() {
+
+  // Capture the class variables and write them as a binary block.
+  struct self_t self;
+  memset(&self, 0, sizeof(self));
```
Its my old C roots that are showing :)
However, initializing with {} does a field by field initialization.  That is not what I am looking for. I want to zero any compiler generated padding and the {} would not do it.  So I think it should remain memset(). I will comment that this structure must contain only POD (plain old data).

breznak replied:
> I want to zero any compiler generated padding and the {} would not do it 

I'm interested why the "padding" bits would matter? As we are, so far, compiler&platform dependant.

dkeeney commented on src/nupic/algorithms/Cells4.cpp:
```diff
@@ -1880,206 +1892,373 @@ void Cells4::reset() {
  * Need to load and re-propagate activities so that we can really persist
  * at any point, load back and resume inference at exactly the same point.
+ *
+ * Note: Most parts are saved in binary format. The file
```
It is not the size that is important it is the enormous time required to format/parse to and from text.  Even for a small implementation the time is significant.

breznak replied:
...understood, and pity

dkeeney commented:
@breznak Thanks for taking time to review this for me.  It is a big help.

dkeeney commented:
Missed updating all of the Py tests in bindings/py/ to remove capnproto.  This is going to take a while.

breznak commented:
Thanks for your work, @dkeeney , this will be a really useful cleanup! 

I've started reading on upcoming parallelism in c++17, so I'm looking forward to that shift afterwards

dkeeney commented:
Thanks for looking at this.
Yes, I am still trying to figure out what is generated and what is hand
coded in the nupic/bindings/py folder.  I would think that generated stuff
should be moved to the build folder so it does not get checked in....or at
least tell git not to include it.

I will probably not get much done until after the laber day weekend...big
family thing coming up.  But I will continue to work on this.



On Wed, Aug 29, 2018 at 2:15 AM breznak <notifications@github.com> wrote:

> *@breznak* requested changes on this pull request.
>
> Did a quick review of the last commit.
> The good:
>
>    - fixed build (capnp stuff)
>    - improved some comments since prev reviews
>    - ArrayAlgo ok
>
> The bad:
>
>    - seems you've accidentally commited stuff in
>    nupic/bindings/*.{so,py} which should not be here (autogenerated)
>    - also one more config (local)
>    TODO:
>    - as you've said, the pytests/bindings need accomodate for the changes
>    too
>
> ------------------------------
>
> In .kdev4/nupic.core.kdev4
> <https://github.com/htm-community/nupic.cpp/pull/62#discussion_r213595935>
> :
>
> > @@ -0,0 +1,40 @@
> +[Buildset]
>
> this is some residual local config, should remove the file(s) before final
> merge
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/htm-community/nupic.cpp/pull/62#pullrequestreview-150462811>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBa_9jzziht6UE0gr0G2Zv4HKHoLAGcks5uVluygaJpZM4WEcXF>
> .
>

breznak commented:
> Yes, I am still trying to figure out what is generated and what is hand coded in the nupic/bindings/py folder 

is your .gitignore in order? Mine does not check-in any of that stuff
```
...
# SWIG
bindings/py/src/nupic/bindings/*.cxx
bindings/py/src/nupic/bindings/*.py
bindings/py/src/nupic/bindings/*.so
bindings/py/src/nupic/*.so
bindings/py/src/nupic.bindings.egg-info
```
I think only the `bindings/*.i` files are relevant. 

Not sure about moving it to build/, anyway, let's leave that for separate PR.. 

Enjoy your time with family!

breznak commented:
> There are lots of tests for capnproto style serialization but nothing for stream serialization so I will have to write some new tests. Have to think how this will tie into pickling of if it needs to.  

If it's pytests for serialization you speak of, maybe we could drop them? And keep mostly only c++ tests for serialization (+ extend them for some extra logic/testcases if found in PY?)

breznak commented:
Btw, 

> The bindings/*.i is under nupic.cpp/src/nupic/bindings so that is a different bindings. 

this is where the manual code is, 

> These in nupic.cpp/bindings/py seem to be where the py tests reside. I had overlooked them entirely.

I think all of this is created by SWIG during build, thus can be ignored

breznak commented on src/nupic/bindings/algorithms.i:
```diff
@@ -1337,31 +1337,31 @@ void forceRetentionOfImageSensorLiteLibrary(void) {
+    //                previously serialized Connections instance.
+
+    //  """
+    //  self._initFromCapnpPyBytes(proto.as_builder().to_bytes()) # copy * 2
```
remove this commented code? If it were ever needed, we have git history

dkeeney replied:
done

breznak commented on bindings/py/src/nupic/bindings/regions/PyRegion.py:
```diff
@@ -324,67 +326,7 @@ def deSerializeExtraData(self, filePath):
-    self.writeToProto(regionImpl)
-
-
   @classmethod
```
remove this line as well

breznak commented on bindings/py/src/nupic/bindings/tools/serialization_test_py_region.py:
```diff
@@ -69,41 +58,8 @@ def randomSeed(self):
-    self._rand.write(proto.random)
-
 
   @classmethod
```
remove this line

dkeeney replied:
No, executeMethod() that follows needs to be declared as a classmethod.  Otherwise I think it would be declared abstract.

breznak replied:
Do you reply to a good line/file? There's no executeMethod here. The decorator belonged to 'readFromProto', and now would fall down to 'initialize'

dkeeney replied:
Hmmm, I must have been looking at the wrong file.  Removed line 62 of serialization_test_py_region.py

breznak commented on bindings/py/tests/nupic_random_test.py:
```diff
@@ -44,21 +36,17 @@
 
 
-  @unittest.skipUnless(
-    capnp, "pycapnp is not installed, skipping serialization test.")
```
keep the skipUnless logic? Could be nice to have a variable serialization (former capnp) and if set to None, the tests are skipped.

breznak commented on src/nupic/bindings/algorithms.i:
```diff
@@ -1337,31 +1337,31 @@ void forceRetentionOfImageSensorLiteLibrary(void) {
 
 
     @classmethod
```
remove/comment out too

breznak commented on src/nupic/regions/PyRegion.cpp:
```diff
@@ -329,8 +329,17 @@ void PyRegion::serialize(BundleIO &bundle) {
-	// @TODO
+	std::ostream &f = bundle.getOutputStream();
+	// cPickle.dump(node_, f, HIGHEST_PROTOCOL)
+	py::Module pickle("pickle")
```
missing ';' ? 
https://travis-ci.org/htm-community/nupic.cpp/builds/422329818#L3876

dkeeney replied:
oops

dkeeney commented:
Hmmm, CI seems to be really broken now.  Don't know what I did that might have caused that.  My company is arriving in a few min so I will be out-of-pocket until Monday afternoon. I may be able to check my email but not much else.

dkeeney commented:
Appveyor is still failing for some unknown reason.  Seems to be confused about the Serializable base class in the SWIG magic someplace.
The other two were successful but to do so I had to comment out the python tests.  The python tests do not seem to be able to load their modules.  The capnproto code had some magic that loaded the path. I cannot figure out how that should happen.  Anyway, all of this will be replaced with pybind11 so I don't think that is a real problem.

dkeeney commented:
After looking at appveyor a little more, it says it has written out the maps and then it errors saying it cannot find the maps.  Don't know what to make of this other than that it is broken.
I have an idea of how I might get the python tests to work in the other two CI's.  I think I will give that at a try.

dkeeney commented:
My idea did not work.  
There are actually two errors that stop the python tests from running. The first is some sort of undefined symbol:  _ZN5nupic6Region9intializeEv when loading nupic.bindings.engine_internal.   Probably something to do with nupic.Region.initialize() but there is nothing unusual about it that I can see.

The second error is an undefined name 'bindings' in class CState(bindings.math.Serializable):  in algorithms.py which is built by SWIG.  Serializable is indeed defined and it is in the math.so.  So I would expect SWIG to reference it as nupic.bindings.math.Serializable.   This is not just CState() but every class that subclasses Serializable.

So I am at a loss as to what to try next.

dkeeney commented:
I tried to update the src/CMakeLists.txt SWIG_ADD_MODULE to SWIG_ADD_LIBRARY.  That got rid of the warning but did not fix the problem.  The .expmap is still not generated under appveyor. The other two CI's have no problem.  Perhaps it is something about Windows wingw that is having a problem with it.

breznak commented:
@dkeeney sorry, I just arrived home from a conference. I'm going to catch my sleep deficit and I'll have a look asap tmr! 

So we're considering breaking the Windows for now..

> The first is some sort of undefined symbol: _ZN5nupic6Region9intializeEv when loading nupic.bindings.engine_internal. Probably something to do with nupic.Region.initialize() but there is nothing unusual about it that I can see. 

There was a hack in capnp for adding a symbol, I remember seeing it being deleted. Maybe we'd have to hack it in.

dkeeney commented:
Found a misspelled class method  Region::intialize() which was not implemented.  Appveyor seemed to have a problem resolving what Swig generated for that situation.

dkeeney commented:
Well all three CI's now fail on the same Python test in the same way.  So Windows CI is working again.
I now need to figure out how to tell Swig how to generate code that will load the Serializable class.

breznak commented:
>  So Windows CI is working again.
I now need to figure out how to tell Swig how to generate code that will load the Serializable class. 

This is good progress, so we have a single problem and even know the cause. Regarding the SWIG - I think we can find the approach from capnp that used Serializabe.

dkeeney commented:
It mostly works :)   One of the Python tests crashed on appveyer. Will have to look at that in the morning.
I had a problem getting Swig to recognize the Serializable class.  I tried everything with no luck.  Ended up overriding the saveToFile() and loadFromFile() functions in the subclasses so it did not need the Serializable class exposed to Python.  

Also, had to comment out one of the Python test which performed serialization of the Python modules. That part not yet implemented. PyRegion.cpp line 293. That can be put off until the Python interface PR.

dkeeney commented:
Well, I thought it might have been lack of permissions to write to a file (for serialization) but after finding a way to obtain a valid tmp directory from py.test it still fails.   :(  Not just a failed test but a hard crash of the test environment.  So I guess I need to install MinGW just to debug this problem.

dkeeney commented:
Finally :)  
I think I have addressed all of the comments so far.
I am going to be out-of-packet for the next few days....one of the advantages of being retired.  I will check email but no more coding until next week.

For next project...perhaps something simple like adding Smart pointers to Region and Link objects.

breznak commented:
Wow, this has been an amazing and a loong way, muchas gracias @dkeeney !! I'll review all of it again today and let's merge this beast :1st_place_medal: Enjoy your rest and bathe in sunshine of passing CI! :)

breznak commented on bindings/py/tests/network_test.py:
```diff
@@ -91,42 +83,43 @@ class NetworkTest(unittest.TestCase):
-      self.assertEqual(destRegion.getParameterUInt32("randomSeed"), 99)
+#########################################
+#Skipping this test for now...
+#TODO: Need to implement pickle serialization of the Python code.
```
this is still missing? Guess left over to pybind11 (?)

breznak commented on nupic.core.kdev4:
```diff
@@ -0,0 +1,4 @@
+[Project]
+CreatedFrom=CMakeLists.txt
+Manager=KDevCMakeManager
```
rm this file, please

breznak replied:
probably you had the file first check to git, and later added to gitignore - it still stays in the repo. Just let the file removed here, and keep your local version - now it should not show in the public git, but you should keep your local config.

breznak commented on src/nupic/algorithms/SpatialPooler.cpp:
```diff
@@ -32,14 +32,15 @@
 #include <nupic/math/Topology.hpp>
-#include <nupic/proto/SpatialPoolerProto.capnp.h>
+
+#define VERSION 2  // version for stream serialization
```
ping

breznak commented:
Merging capnp removal, replaced by native bitstream serialization! 
Huge work done by @dkeeney , thank you very much!! :clap: :clap:

breznak commented on .gitignore:
```diff
@@ -17,19 +18,29 @@ htmlcov/
 
+# KDevelop
+.kdev4
+nupic.core.kdev4
```
sorry, I just deleted that. Best solution would be to have the IDE to keep the files outside the git project's folder/scope, so individual settings don't have to be commited. Or just add it next time please.
